### PR TITLE
UI: Expose sketch attributes through a dynamic table display component

### PR DESCRIPTION
**Pull request by tomchop**
## Open questions & comments for reviewers

- Where should this format be documented?
- Is the current architecture fine (i.e. DynamicTableDisplay component in `/Common/` as it will probably be reused)?
- More advanced features for this display (e.g. buttons and callbacks that mutate rows) should be pushed in a future PR as I feel this one will already spawn quite a lot of discussion.
- Happy to write tests for this once we no longer have comments on the format / architecture 😄 

## Pull request details

+ What existing problem does this PR solve?

There is currently no way to display complex attributes attributes (e.g. objects) in a way that is user-friendly.  Fixes https://github.com/google/timesketch/issues/1797

+ What new feature is being introduced with this PR?

This PR introduces a DynamicTableDisplay to display attributes in a user-friendly table. The way the tables are formatted (e.g. column names and labels) are defined through the attribute itself, in a `columns` field. This component takes two props, a `section` prop and a `data` prop. They must be formatted as follows:

```javascript
let section = {
        key: 'vt_summary',
        label: 'Virustotal Summary',
        columns: [
          { field: 'hash', label: 'SHA256 hash' },
          { field: 'first_seen', label: 'First seen' },
          { field: 'last_seen', label: 'Last seen' },
          { field: 'last_analysis', label: 'Last analysis' },
          { field: 'ratio', label: 'Detection ratio' },
          { field: 'tag', label: 'Tags' }
        ]
      },
```

```javascript
let data = [
      {
        hash: '3ca44cadd76585cd5e61df50ff1bd9276636ad1ee7026e8fd0b7a4a371e70351',
        first_seen: '2014-02-14 16:46:37',
        last_seen: '2020-07-04 14:15:54',
        last_analysis: '2020-07-04 14:15:54',
        ratio: 0.14,
        tag: ['pdf', 'autoaction', 'js-embedded']
      },
      {
        hash: '67451b0a853884dfa134654421603f41a05f8f402659363bc5b950b602af2e1d',
        first_seen: '2013-06-29 10:02:19',
        last_seen: '2021-01-08 17:32:43',
        last_analysis: '2021-01-08 17:32:43',
        ratio: 0.16,
        tag: ['pdf', 'js-embedded']
      },
      {
        hash: '0833778a108c33042374dfc501f03e028d935ddb0dfba7a552f661716b61209e',
        first_seen: '2012-07-14 10:58:23',
        last_seen: '2020-07-04 14:15:52',
        last_analysis: '2020-07-04 14:15:52',
        ratio: 0.16,
        tag: ['pdf', 'js-embedded']
      }
    ]
```

For an example of how to use this component, take a look at the `Intelligence.vue` component.

**Checks**

- [ ] All tests succeed.
- [ ] Unit tests added.
- [ ] e2e tests added.
- [ ] Documentation updated.

## Preview screenshot

image

### CODE REVIEW

File: timesketch/frontend/src/views/Intelligence.vue [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+     v-bind:key="section.key"
+     :section="section"
+     :data="intelligence.data[section.key]">
+    </dynamic-table-display>
+
+  </div>
+</template>
+
+<script>
+import DynamicTableDisplay from '../components/Common/DynamicTableDisplay'
```
Comment (berggren) [inline timesketch/frontend/src/views/Intelligence.vue:40]: We already have a DynamicForm component, maybe stick with the `DynamicTable` for this one to stay a bit consistent in the naming?

I also use a `ts-` prefix for Timesketch internal components.

```suggestion
import TsDynamicTable from '../components/Common/DynamicTable'
```

File: timesketch/frontend/src/views/Intelligence.vue [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      currentAppContext="sketch"
+      currentPage="intelligence"
+    ></ts-navbar-secondary>
+
+    <dynamic-table-display
+     v-for="section in intelligence.meta.sections"
+     v-bind:key="section.key"
+     :section="section"
+     :data="intelligence.data[section.key]">
+    </dynamic-table-display>
```
Comment (berggren) [inline timesketch/frontend/src/views/Intelligence.vue:34]: see rational in a comment later in this file

```suggestion
    <ts-dynamic-table
     v-for="section in intelligence.meta.sections"
     v-bind:key="section.key"
     :section="section"
     :data="intelligence.data[section.key]">
    </ts-dynamic-table>
```

File: timesketch/frontend/src/components/Common/DynamicTableDisplay.vue [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        </div>
+      </div>
+    </section>
+
+  </div>
+</template>
+
+<script>
+export default {
+  name: 'DynamicTableDisplay',
```
Comment (berggren) [inline timesketch/frontend/src/components/Common/DynamicTableDisplay.vue:60]: IMO Display in the component name is not needed. If you approve, also change the filename.

```suggestion
  name: 'DynamicTable',
```

File: timesketch/frontend/src/components/Common/DynamicTableDisplay.vue [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+<template>
+  <div>
+    <section class="section">
+      <div class="container is-fluid">
+        <div class="card" style="min-height:160px;">
```
Comment (berggren) [inline timesketch/frontend/src/components/Common/DynamicTableDisplay.vue:20]: Do we want to include the wrapping element/style?

To make the component a bit easier to use in different contexts you could just render the table and let the caller (view) add it to whatever element (card, section etc) it wants.

WDYT?

**Review by berggren — CHANGES_REQUESTED** [review]
Looks good. Some suggestions to consider.

**Comment by berggren** [comment]
> ## Open questions & comments for reviewers
> * Where should this format be documented?

Good question. Initially we could just document it in the DynamicTable.vue file?

> * Is the current architecture fine (i.e. DynamicTableDisplay component in `/Common/` as it will probably be reused)?

Yeah, /Common is a good home for this component.

> * More advanced features for this display (e.g. buttons and callbacks that mutate rows) should be pushed in a future PR as I feel this one will already spawn quite a lot of discussion.

Ack, I agree! :)

> * Happy to write tests for this once we no longer have comments on the format / architecture 😄

Yay! ;)

> 
> ## Pull request details
> * What existing problem does this PR solve?
> 
> There is currently no way to display complex attributes attributes (e.g. objects) in a way that is user-friendly. Fixes #1797
> 
> * What new feature is being introduced with this PR?
> 
> This PR introduces a DynamicTableDisplay to display attributes in a user-friendly table. The way the tables are formatted (e.g. column names and labels) are defined through the attribute itself, in a `columns` field. This component takes two props, a `section` prop and a `data` prop. They must be formatted as follows:
> 
> ```js
> let section = {
>         key: 'vt_summary',
>         label: 'Virustotal Summary',
>         columns: [
>           { field: 'hash', label: 'SHA256 hash' },
>           { field: 'first_seen', label: 'First seen' },
>           { field: 'last_seen', label: 'Last seen' },
>           { field: 'last_analysis', label: 'Last analysis' },
>           { field: 'ratio', label: 'Detection ratio' },
>           { field: 'tag', label: 'Tags' }
>         ]
>       },
> ```
> 
> ```js
> let data = [
>       {
>         hash: '3ca44cadd76585cd5e61df50ff1bd9276636ad1ee7026e8fd0b7a4a371e70351',
>         first_seen: '2014-02-14 16:46:37',
>         last_seen: '2020-07-04 14:15:54',
>         last_analysis: '2020-07-04 14:15:54',
>         ratio: 0.14,
>         tag: ['pdf', 'autoaction', 'js-embedded']
>       },
>       {
>         hash: '67451b0a853884dfa134654421603f41a05f8f402659363bc5b950b602af2e1d',
>         first_seen: '2013-06-29 10:02:19',
>         last_seen: '2021-01-08 17:32:43',
>         last_analysis: '2021-01-08 17:32:43',
>         ratio: 0.16,
>         tag: ['pdf', 'js-embedded']
>       },
>       {
>         hash: '0833778a108c33042374dfc501f03e028d935ddb0dfba7a552f661716b61209e',
>         first_seen: '2012-07-14 10:58:23',
>         last_seen: '2020-07-04 14:15:52',
>         last_analysis: '2020-07-04 14:15:52',
>         ratio: 0.16,
>         tag: ['pdf', 'js-embedded']
>       }
>     ]
> ```
> 
> For an example of how to use this component, take a look at the `Intelligence.vue` component.
> 
> **Checks**
> 
> * [ ]  All tests succeed.
> * [ ]  Unit tests added.
> * [ ]  e2e tests added.
> * [ ]  Documentation updated.
> 
> ## Preview screenshot
> image

File: timesketch/frontend/src/components/Common/DynamicTableDisplay.vue (hunk 4 above)
Comment (tomchop) [reply timesketch/frontend/src/components/Common/DynamicTableDisplay.vue]: My idea here was to encourage some consistency on the UI. The wrapping is fairly minimal here; I think most users of that component would be OK with it.

**Review by berggren — APPROVED** [review]

**Pull request merged**